File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-2.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l = 0
        r = len(nums) - 1

        while l < r:
            m = l + ((r - l)//2)

            if nums[m] >= nums[r]:
                l = m + 1
            elif nums[m] <= nums[r]:
                r = m 
        
        left = 0 
        right = len(nums) - 1

        if target >= nums[l] and target <= nums[right]:
            left = l
        else:
            right = l - 1  
        
        while left <= right:
            mid = left + ((right - left) // 2)

            if target < nums[mid]:
                right = mid - 1
            elif target > nums[mid]:
                left = mid + 1
            else:
                return mid
            
        return -1

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    // -1000 <= Node.val <= 1000
    int maxi = -1001;

    public int maxPathSum(TreeNode root) {
        dfs(root);

        return maxi;
    }

    public int dfs(TreeNode node)
    {
        if (node == null) return 0;

        int left = dfs(node.left);
        int right = dfs(node.right);

        maxi = Math.max(maxi, left + right + node.val);

        return Math.max(0, node.val + Math.max(left, right));
    }
}
